### PR TITLE
Fix issue for building docker container

### PR DESCRIPTION
**Pull request by luclaurent**
Many improvements and updates on
 - workflow (update versions and missing id
 - Dockerfile to use new setup version and new testing

### CODE REVIEW

File: .github/workflows/build-container.yml [hunk 1]
```diff
@@ -13,24 +13,25 @@ jobs:
     name: 'Build amc2moodle container'
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v4
```
Comment (nennigb) [inline .github/workflows/build-container.yml:16]: I suggest `@main` instead of `@v4` for all actions.
Since the version doesn't matter, `@main ` allow to use always the last version.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "python.testing.unittestArgs": [
+        "-v",
+        "-s",
+        "./amc2moodle/tests",
+        "-p",
+        "test_*.py"
+    ],
+    "python.testing.pytestEnabled": false,
```
Comment (nennigb) [inline .vscode/settings.json:9]: Is this file really required ?

File: amc2moodle/utils/misc.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    cls : class
+        return the decorated class.
+    """
+    def decorate(cls):
+        for name, fn in inspect.getmembers(cls, inspect.ismethod):
+            setattr(cls, name, decorator(fn))
+        return cls
+    return decorate
+
+def decorator_set_cwd(directory):
```
Comment (nennigb) [inline amc2moodle/utils/misc.py:80]: Perhaps there are more pythonic way to do this ?
Defined in unittest famework ?

File: .vscode/settings.json (hunk 2 above)
Comment (luclaurent) [reply .vscode/settings.json:9]: :-D probably not witht the future new version using pyproject.toml (vscode will read it to identify tests)

File: amc2moodle/utils/misc.py (hunk 3 above)
Comment (luclaurent) [reply amc2moodle/utils/misc.py:80]: I don't know but it works well :-)

File: .vscode/settings.json (hunk 2 above)
Comment (luclaurent) [reply .vscode/settings.json:9]: I removed it in branch new-setup-hatchling

**Review by nennigb — CHANGES_REQUESTED** [review]
I found also few 'problems'.

  - [ ] I get  'docker permission denied mount volume', I need for instance ` --user 1000:1000`  to write in `/tmp/work`. Perhaps it should be mentioned at the end of README. Or or instead of the first 1000, `$(id -u)`
  - [ ]  Perhaps also add that DIRA and DIRB should be replaced by true path. Close to line 23 in the README. It took me time to see my mistake due to copy/paste

**Comment by nennigb** [comment]
Another thing to fix (one day) would be the log that grows indefinitely if the docker is not stopped.
In my opinion it is not a big issue since it not the heart of the project.

File: docker/Dockerfile [hunk 4]
```diff
@@ -3,39 +3,47 @@ FROM texlive/texlive:latest
...
+    ghostscript \
+    imagemagick \
+    libtext-unidecode-perl \
+    latexml \
+    xmlindent \
+    python3-pip \
+    python3-venv \
+    git && \
+    apt clean &&\
+    rm -rf /var/lib/{apt,dpkg,cache,log}/
```
[the remark this answers is not in the document]
Comment (luclaurent) [reply docker/Dockerfile:28]: fixed

File: docker/Dockerfile [hunk 5]
```diff
@@ -3,39 +3,47 @@ FROM texlive/texlive:latest
...
 # install pip and Python pkg
-# RUN pip3 install -U pip && \
-#     pip install amc2moodle
 WORKDIR /tmp
 RUN git clone https://github.com/nennigb/amc2moodle.git -b master ${INSTALL_DIR_A2M}
 WORKDIR ${INSTALL_DIR_A2M}
-RUN pip3 install -U pip && \
-    pip3 install .
+RUN mkdir -p ${VENVDIR} && \
+    python -m venv ${VENVDIR}
```
[the remark this answers is not in the document]
Comment (luclaurent) [reply docker/Dockerfile:37]: in order to avoid "error: externally-managed-environment" when use the system python/pip

File: docker/Dockerfile [hunk 6]
```diff
@@ -9,11 +9,22 @@ LABEL org.opencontainers.image.source = "https://github.com/nennigb/amc2moodle/"
...
+    latexml \
+    xmlindent \
+    python3-pip \
+    python3-venv \
+    git && \
+    apt clean &&\
+    rm -rf /var/lib/{apt,dpkg,cache,log}/
 
 # move policy file
 RUN mv /etc/ImageMagick-6/policy.xml /etc/ImageMagick-6/policy.xml.off
```
[the remark this answers is not in the document]
Comment (luclaurent) [reply docker/Dockerfile:30]: fixed

File: docker/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (nennigb) [reply docker/Dockerfile:37]: ok  :-)

File: pyproject.toml [hunk 7]
```diff
@@ -26,6 +26,10 @@ dependencies = [
 ]
 requires-python='>=3.8'
 
+[project.optional-dependencies]
+test = ["pytest", "pytest-cov[all]"]
+lint = [ "black", "flake8"]
```
[the remark this answers is not in the document]
Comment (luclaurent) [reply pyproject.toml:31]: I add it

**Review by nennigb — APPROVED** [review]
For me it goods. You can have a look on the detailed the comments.
I let you merge when you are ready.

**Comment by nennigb** [comment]
Just a question the modification with f strings were automatic or manual ?
I checked most of them but just in case.

**Comment by luclaurent** [comment]
> Just a question the modification with f strings were automatic or manual ? I checked most of them but just in case.

I ran `hatch fmt` by mistake. The correct syntax to only check is `--check`. But after that, I kept the changes ...

File: .github/workflows/build-container.yml (hunk 1 above)
Comment (luclaurent) [reply .github/workflows/build-container.yml]: Caution @main is not always available (check for @master)
